iroh-gossip/src/proto/topic.rs

```diff
@@ -0,0 +1,858 @@
+    pub fn with_rng(me: PA, me_data: PeerData, config: Config, rng: R) -> Self {
+        Self {
+            swarm: hyparview::State::new(me, me_data, config.membership, rng),
+            gossip: plumtree::State::new(me, config.broadcast),
+            me,
+            outbox: VecDeque::new(),
+            stats: Stats::default(),
+        }
+    }
+
+    /// Handle an incoming event.
+    ///
+    /// Returns an iterator of outgoing events that must be processed by the application.
+    pub fn handle(
+        &mut self,
+        event: InEvent<PA>,
+        now: Instant,
+    ) -> impl Iterator<Item = OutEvent<PA>> + '_ {
+        let io = &mut self.outbox;
+        // Process the event, store out events in outbox.
+        match event {
+            InEvent::Command(command) => match command {
+                Command::Join(peer) => self.swarm.handle(SwarmIn::RequestJoin(peer), now, io),
+                Command::Broadcast(data) => self.gossip.handle(GossipIn::Broadcast(data), io),
+            },
+            InEvent::RecvMessage(from, message) => {
+                self.stats.messages_received += 1;
+                match message {
+                    Message::Swarm(message) => {
+                        self.swarm
+                            .handle(SwarmIn::RecvMessage(from, message), now, io)
+                    }
+                    Message::Gossip(message) => {
+                        self.gossip.handle(GossipIn::RecvMessage(from, message), io)
+                    }
+                }
+            }
+            InEvent::TimerExpired(timer) => match timer {
+                Timer::Swarm(timer) => self.swarm.handle(SwarmIn::TimerExpired(timer), now, io),
+                Timer::Gossip(timer) => self.gossip.handle(GossipIn::TimerExpired(timer), io),
+            },
+            InEvent::PeerDisconnected(peer) => {
+                self.swarm.handle(SwarmIn::PeerDisconnected(peer), now, io);
+                self.gossip.handle(GossipIn::PeerDisconnected(peer), io);
+            }
+            InEvent::UpdatePeerData(data) => {
+                self.swarm.handle(SwarmIn::UpdatePeerData(data), now, io)
+            }
+        }
+
+        // Forward NeigborUp and NeighborDown events from hyparview to plumtree
+        let mut io = VecDeque::new();
+        for event in self.outbox.iter() {
+            match event {
+                OutEvent::EmitEvent(Event::NeighborUp(peer)) => {
+                    self.gossip.handle(GossipIn::NeighborUp(*peer), &mut io)
+                }
+                OutEvent::EmitEvent(Event::NeighborDown(peer)) => {
+                    self.gossip.handle(GossipIn::NeighborDown(*peer), &mut io)
+                }
+                _ => {}
+            }
+        }
+        // Note that this is a no-op because plumtree::handle(NeighborUp | NeighborDown)
+        // above does not emit any OutEvents.
+        self.outbox.extend(io.drain(..));
```

Comment:
why not make `Outbox` be a `VecDeque` and pass it directly through?

Reply:
`self.outbox` is a `Vec` but could be a `VecDeque` just the same - the issue is that in the lines above, I'm iterating over `self.outbox` and then padding references of some events into the `self.gossip.handle`, therefore I cannot pass `&mut self.outbox` into there as well (because that would be a mutable ref plus an immutable ref, which of course cannot work).